### PR TITLE
Check Hooks

### PR DESCRIPTION
**Pull request by portertech**
The Sensu client’s primary function is to execute check commands and publish their results for processing, allowing a Sensu server to take the appropriate actions. The practice of using Sensu for automated data gathering for incident triage continues to gain traction amongst Sensu users. Although this practice is achievable, Sensu does not currently make it easy, as all actions must be orchestrated by a Sensu server.

This pull request introduce check hooks, commands run by the Sensu client in response to the result of the check command execution. The Sensu client will execute the appropriate configured hook command, depending on the check execution status (e.g. 1). Valid hook names include (in order of precedence): "1"-"255", "ok", "warning", "critical", "unknown", and "non-zero". The check hook command output, status, executed timestamp, and duration are captured and published in the check result.

Example check hook configuration:

``` json
{
  "checks": {
    "my_check": {
      "command": "echo WARNING && exit 1",
      "...": "...",
      "hooks": {
        "ok": "echo ALL IS WELL",
        "warning": "echo DANGER WILL ROBINSON!",
        "critical": "ps aux"
      }
    }
  }
}
```

Example check result hook data:

```
{
  "hooks": {
    "warning": {
      "command": "echo HOOKED",
      "timeout": 5,
      "output": "HOOKED\n",
      "status": 0,
      "executed": 1506049908,
      "duration": 0.004
    }
  }
}
```

Closes https://github.com/sensu/sensu/issues/1680

### CODE REVIEW

**Comment by portertech** [comment]
Now to get feedback from users 🎉 

YARDOC and sensu-settings validation to come tomorrow.

**Comment by portertech** [comment]
For posterity, I did consider using an array of hook commands, e.g. `"warning": [{"command": "..."}, {"command": "..."}]`, in order to support multiple sequential executions (if the previous exited 0), however, it adds unnecessary complexity as the same can be achieved with existing (and more powerful) shell operators, e.g. `&&` and `||`.

**Comment by majormoses** [comment]
@portertech agreed that really we can just rely on shell/ruby operators to accomplish different that. 

Regarding different hook types when we discussed this (yes a while ago) I could think of two basic types of hooks: command and http(s). Are you thinking that we would just write a ruby, shell, python, etc script to retrieve data (such as $HOST:PORT/metrics) or should we have a bit more native approach?

If so I was thinking something along these lines:
```json
{
  "checks": {
    "my_check": {
     "hook_type" : "(command|http)"
      "hook_exec": "echo WARNING && exit 1",
      "...": "...",
      "hooks": {
        "ok": "echo ALL IS WELL",
        "warning": "echo DANGER WILL ROBINSON!",
        "critical": "ps aux"
      }
    }
  }
}
```

Basically just whether I need to either write wrapper scripts, use `curl $HOST:$PORT/$PATH`, or if we should support an HTTP endpoint natively.

**Comment by majormoses** [comment]
To be clear I am not saying we should support it natively just asking the question. An argument to be made against it is that now we basically have to support every header option under the sun as [someone will ask for it](https://github.com/hashicorp/consul/issues/1184).

**Comment by portertech** [comment]
@majormoses I am indeed thinking that we will write and use shell, ruby, golang, python, etc scripts to retrieve data and do things 👍

**Comment by majormoses** [comment]
> "echo DANGER WILL ROBINSON!"

I am not gonna lie I kinda want to add this to all my monitoring now.

File: spec/config.json [hunk 1]
```diff
@@ -196,6 +196,19 @@
...
+      "command": "echo FAIL && exit 1",
+      "subscribers": [
+        "test"
+      ],
+      "hooks": {
+          "warning": {
+              "command": "echo HOOKED",
+              "timeout": 5
+          }
+      },
```
Comment (davideauee) [inline spec/config.json]: Note that the commit message is describing a different config format.

Also, a default timeout should probably be enforced or the attribute made mandatory.
If "Spawn.process" does already enforce something then it should probably be mentioned in the comments/docs

File: lib/sensu/client/process.rb [hunk 2]
```diff
@@ -110,6 +110,46 @@ def check_in_progress_key(check)
...
+      #   hook (if any).
+      def execute_check_hook(check)
+        @logger.debug("attempting to execute check hook", :check => check)
+        severity = SEVERITIES[check[:status]] || "unknown"
+        hook = check[:hooks][check[:status].to_s] || check[:hooks][severity.to_sym]
+        if hook.nil? && check[:status] != 0
+          hook = check[:hooks]["non-zero"]
+        end
+        if hook
+          command, unmatched_tokens = substitute_tokens(hook[:command], @settings[:client])
```
Comment (davideauee) [inline lib/sensu/client/process.rb]: in addition to leveraging client config tokens, it could be convenient to provide/document a mechanism to pass all or parts of check result event to the hook's command as well.

Comment (calebhailey) [reply lib/sensu/client/process.rb]: Oh, that's a good idea.

Comment (davideauee) [reply lib/sensu/client/process.rb]: Glad to hear 😄,

...  because another thought that I cannot quite shake is that this is new feature is bound to be used to cannabelize server side handlers/extensions. For example, an easy temptation would be to use those hooks to fanout processing logic that can get problematic when funneled through a reduced number of sensu server instances.

I fully understand that what follows is quite a stretch from the intent of this pull request but... 

In effect what we have here is a poor's man client side handler but without the supporting framework that server side handlers/extensions enjoy (configuration, logging, defined interface, etc.).

And if we follow this train of thought to it logical conclusion... should this hook concept really be about allowing the client to run handlers/extension to process check result payloads?

It would arguably be a heavier process than the proposed desgin (at the very least you would likely be bound to use ruby) but it would also be more powerfully integrated, more inlined with the existing patterns, would not require to specify a separate command/handler for each status value, etc...

I feel a bit like the devil's advocate here but still,... food for thoughts.

d.

Comment (portertech) [reply lib/sensu/client/process.rb]: @davideauee thank you for the feedback!

I could add `"stdin": true` as a hook option (default to `false`), and have the Sensu client write the check definition with the populated `output`, `status`, etc to the hook process STDIN 👍 

Providing "semi event data" to the hook process STDIN is a possibility, however, it would be lacking the context constructed by the Sensu server, e.g. history, flapping, etc, and functionality like silencing would not work. I personally do not like the idea of overloading Check Hooks.

File: spec/config.json (hunk 1 above)
Comment (portertech) [reply spec/config.json]: We can introduce a default timeout for hooks, as it will not break backwards compatibility 👍

File: lib/sensu/client/process.rb (hunk 2 above)
Comment (davideauee) [reply lib/sensu/client/process.rb]: leveraging STDIN would do the trick indeed 👍 .

I see what you mean by not wanting to "overload" Check Hooks as the intent here is to provide something both light weight and pragmatic, but it arguably comes at a cost of some amount of shoehorning.   

looking at the pros of each approach:
- client-side handlers:
   - more consistent with already established and well understood patterns 
   - well defined interfaces to feed data (configs, event data, etc.) so no need to shoehorn it through more esoteric means (config tokens, split triggers based on status, stdin, etc.)
   - could have its own config
   - better at promoting reuse
- check hooks:
   - not limited to ruby
   - easier to leverage existing local admin/ops tools (no need for a ruby shim) 
   - should be faster (at the very least, no need to bring up a ruby vm)

File: spec/config.json (hunk 1 above)
Comment (portertech) [reply spec/config.json]: @davideauee update the PR to set/use a default hook timeout of 60 seconds.

**Comment by portertech** [comment]
@davideauee I have update the PR with hook "stdin" support, an option to write JSON serialized Sensu client and check data to the hook command process via STDIN. The JSON data closely resembles the Sensu event data that goes through the event pipeline on Sensu servers. I avoided any reference to "events", as check hooks are not intended to be client side event handlers.

**Comment by portertech** [comment]
We have an opportunity to add "stdin" to checks, providing optional client and check data to check plugins.

**Comment by portertech** [comment]
50-caliber-rambo-kitty-assault-unicorn

**Comment by MrMoronIV** [comment]
So this is implemented now and this is the way to define them?


{
  "checks": {
    "my_check": {
      "command": "echo WARNING && exit 1",
      "...": "...",
      "hooks": {
        "ok": {"command": "echo ALL IS WELL"},
        "warning": {"command": "echo DANGER WILL ROBINSON!"},
        "critical": {"command": "ps aux"}
      }
    }
  }
}

**Comment by majormoses** [comment]
Its not in the example docs (would love a PR to update them:
```
{
  "checks": {
    "my_check": {
      "command": "echo WARNING && exit 1",
      "...": "...",
      "hooks": {
        "ok": {
          "command": "echo ALL IS WELL",
          "timeout": 1
        },
        "warning": {
          "command": "echo DANGER WILL ROBINSON!",
          "timeout": 5
          },
        "critical": {
          "command": "ps aux",
          "timeout": 10
          }
      }
    }
  }
}```
